src/Simplify/AstHelpers.elm

```diff
@@ -950,6 +952,74 @@ isSpecificUnappliedBinaryOperation symbol checkInfo expression =
             False
 
 
+canEqualOrContainNaN : Node Expression -> Bool
+canEqualOrContainNaN node =
+    canEqualOrContainNaNHelp [ node ]
+
+
+canEqualOrContainNaNHelp : List (Node Expression) -> Bool
+canEqualOrContainNaNHelp nodes =
+    case nodes of
+        first :: rest ->
+            case Node.value first of
+                Expression.IfBlock condition thenBranch elseBranch ->
+                    canEqualOrContainNaNHelp (condition :: thenBranch :: elseBranch :: rest)
+
+                Expression.TupledExpression newNodes ->
+                    canEqualOrContainNaNHelp (newNodes ++ rest)
+
+                Expression.ParenthesizedExpression newNode ->
+                    canEqualOrContainNaNHelp (newNode :: rest)
+
+                Expression.ListExpr newNodes ->
+                    canEqualOrContainNaNHelp (newNodes ++ rest)
+
+                Expression.Application _ ->
+                    True
+
+                Expression.OperatorApplication operator _ _ _ ->
+                    -- If the operator can lead to a number being returned, then it's possible the expression
+                    -- evaluates to NaN.
+                    if
+                        (operator == "+")
+                            || (operator == "-")
+                            || (operator == "*")
+                            || (operator == "/")
+                            || (operator == "^")
+                    then
+                        True
+
+                    else
+                        canEqualOrContainNaNHelp rest
```

Comment:
for one, the left and right expressions should be added to the recursive call. For 2, operators `|>` and `<|` can produce NaN